packages/nuxt/src/core/app.ts

```diff
@@ -49,12 +55,23 @@ export async function generateApp (nuxt: Nuxt, app: NuxtApp, options: { filter?:
         nuxt.vfs[fullPath.replace(/\//g, '\\')] = contents
       }
 
+      const perf = performance.measure(fullPath, mark?.name) // TODO: remove when Node 14 reaches EOL
+      const setupTime = perf ? Math.round((perf.duration * 100)) / 100 : 0 // TODO: remove when Node 14 reaches EOL
+
+      if (nuxt.options.debug || setupTime > 500) {
+        console.info(`[nuxt] compiled \`${template.filename}\` in ${setupTime}ms`)
+      }
+
       if (template.write) {
-        await fsp.mkdir(dirname(fullPath), { recursive: true })
-        await fsp.writeFile(fullPath, contents, 'utf8')
+        writes.push(() => {
+          mkdirSync(dirname(fullPath), { recursive: true })
+          writeFileSync(fullPath, contents, 'utf8')
+        })
       }
     }))
 
+  for (const write of writes) { write() }
```

Comment:
AFAIK, using sync methods, blocks (our) main js render loop but it won't speed up v8 i/o threads. Have you tried with this new batching technique ++ async `Promise.all` to see timing differences?

Reply:
I've done some tests but the performance benefit I'm looking for here is reducing cascading hmr.

Reply:
Yes, I understand that main point 👍🏼 But i suppose (native) HMR watchers are not in the same event loop of main javascript process we are writing files from (unless our receiver is sync again). If either method is faster, it reduces final chance to have duplicate HMR events but if you already tried both, feel free to ignore this comment :)

Reply:
it may be an unnecessary optimisation. I tried both without appreciable time difference, and flagged up the fact that we are actually not watching the `buildDir` with vite. I will have a look to see if we actually _should_ be watching the directory. Not doing so might be causing some occasionally issues with vite-node cache, for example.

Reply:
I think you have a good point here and we should definitely iterate on this. I'd be very open to rethinking this if you notice any issue..